firedrake/slate/slate.py

```diff
@@ -0,0 +1,616 @@
+
+        # Generate ufl.Integral data from input form
+        self.generate_integral_info(form.integrals())
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.form.arguments()
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.form.coefficients()
+
+    def __str__(self, prec=None):
+        """String representation of a tensor object in SLATE."""
+        if self.rank == 0:
+            return "S_%d" % self.id
+        elif self.rank == 1:
+            return "V_%d" % self.id
+        elif self.rank == 2:
+            return "M_%d" % self.id
+
+    def __repr__(self):
+        """SLATE representation of the tensor object."""
+        if self.rank == 0:
+            return "Scalar(%r)" % self.form
+        elif self.rank == 1:
+            return "Vector(%r)" % self.form
+        elif self.rank == 2:
+            return "Matrix(%r)" % self.form
+
+
+class UnaryOp(TensorBase):
+    """An abstract SLATE class for representing unary operations on a
+    Tensor object.
+
+    The currently supported unary operations on a SLATE :class:`TensorBase`
+    expression are the following:
+
+        (1) the inverse of a tensor, `A.inv`, implemented in the subclass
+            `Inverse`;
+        (2) the tranpose of a tensor, `A.T`, implemented in the subclass
+            `Transpose`;
+        (3) and the negative operation, `-A` (subclass `Negative`).
+
+    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
+            (:class:`Tensor`) or any derived expression resulting from any number
+            of linear algebra operations on `Tensor` objects. For example,
+            another instance of a `UnaryOp` object is an acceptable input, or
+            a `BinaryOp` object.
+    """
+
+    def __init__(self, A):
+        """Constructor for the UnaryOp class."""
+        if not isinstance(A, TensorBase):
+            raise Exception("Expecting a `slate.TensorBase` object, not %r" % A)
+        self.tensor = A
+        self.check_operand_inversion(A)
+        super(UnaryOp, self).__init__()
+
+    def arguments(self):
+        """Returns a tuple of arguments associated with the tensor."""
+        return self.get_uop_arguments(self.tensor)
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.tensor.coefficients()
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor.
+        """
+        return self.tensor.ufl_domains()
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        `{domain:{integral_type: subdomain_data}}`.
+        """
+        return self.tensor.subdomain_data()
+
+    @classmethod
+    def check_operand_inversion(cls, A):
+        pass
+
+    @classmethod
+    def get_uop_arguments(cls, A):
```

Comment:
Basically, yes. This will also move the method hierarchy: instead of having a generic `arguments` that calls a polymorphic `get_bop_arguments` that have many implementations, you would have a polymorphic `arguments` method with several implementations.